### PR TITLE
Chaos Workflow to validate data consistency and availability

### PR DESCRIPTION
**Pull request by gildaf**
This PR has two parts (two commits):
1. create a small application called ngc-monkey that enables sending traffic to redis database and later check that all keys that were sent sent still has the expected value after the experiment was run
2. create a workflow that installs the app, and then run a delete-pod experiment. The experiment has two probes:
    a. starts the traffic
    b. stop the traffic and check the data

### CODE REVIEW

File: ngc-monkey/Dockerfile.server [hunk 1]
```diff
@@ -0,0 +1,9 @@
+FROM python:3.11.1-buster
+COPY requirements.txt requirements.txt
+RUN pip install -r requirements.txt
+
+COPY ./ ./ngc-monkey
```
Comment (nonirosenfeldredis) [inline ngc-monkey/Dockerfile.server:5]: maybe call it "ngc-monkey-probe"?

File: ngc-monkey/ngc-monkey/server.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    # TODO check if the host already has a running task
+    tg = TrafficGenerator(host=db_name)
+    try:
+        tg.redis_con.assert_raft_is_up()
+    except Exception as e:
+        flask.abort(500, f"Failed to connect to redis {db_name}. {str(e)}")
+
+    tg.start()
+    _threads[db_name] = tg
+    return flask.jsonify(200)
```
Comment (nonirosenfeldredis) [inline ngc-monkey/ngc-monkey/server.py]: I would return 201/202 ( accepted - but in progress & done as in 200)

File: ngc-monkey/ngc-monkey/server.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+def run():
+    """
+    Start a task (spawn a thread) that pushes keys
+    to redis and remember each key that was successfully set.
+    :param: db_name (String) the host name of the Redis Database
+    :return: 200 on success, 500 on failure
+    """
+    db_name = request.json.get("db_name")
+    # TODO check if the host already has a running task
+    tg = TrafficGenerator(host=db_name)
```
Comment (nonirosenfeldredis) [inline ngc-monkey/ngc-monkey/server.py:27]: I would maybe add some defence that if db name is already "in traffic" dont spawn the thread and return some error code

File: ngc-monkey/ngc-monkey/traffic_generator.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        self.result: typing.Optional[Result] = None
+        self.logger = None
+
+    def stop_and_check(self):
+        """
+        Tell the traffic generator to stop sending traffic and check data integrity
+        """
+        self._event.set()
+
+    def _check(self) -> Result:
```
Comment (nonirosenfeldredis) [inline ngc-monkey/ngc-monkey/traffic_generator.py]: is there an issue with threads visibility or atomicity here? when u set the event from one thread ( the http thread ) and u check in the generator thread?

File: ngc-monkey/Dockerfile.server (hunk 1 above)
Comment (gildaf) [reply ngc-monkey/Dockerfile.server:5]: I dont really like it..... 
This isnt a probe. the probe (which is part of the experiment) is using the this app

File: ngc-monkey/ngc-monkey/traffic_generator.py (hunk 4 above)
Comment (gildaf) [reply ngc-monkey/ngc-monkey/traffic_generator.py]: fixed

File: ngc-monkey/ngc-monkey/server.py (hunk 2 above)
Comment (gildaf) [reply ngc-monkey/ngc-monkey/server.py]: fixed

File: ngc-monkey/ngc-monkey/server.py (hunk 3 above)
Comment (gildaf) [reply ngc-monkey/ngc-monkey/server.py:27]: fixed

**Review by nonirosenfeldredis — APPROVED** [review]

**Pull request merged**